src/package/mod.rs

```diff
@@ -0,0 +1,560 @@
+    ///
+    /// This function explores the NPF by extracting it in a temporary folder.
+    pub fn fill_with<P: AsRef<Path>>(&mut self, npf: P) -> Result<(), Error> {
+        let npf_file = File::open(&npf)?;
+        let mut archive = Archive::new(npf_file);
+        let tmp_extract_path = gen_tmp_filename();
+
+        fs::create_dir(&tmp_extract_path)?;
+
+        let res: Result<_, Error> = try {
+            archive.unpack(&tmp_extract_path)?;
+
+            // Bunch of files to analyse/generate
+            let tmp_manifest_path = PathBuf::from(&tmp_extract_path).join("manifest.toml");
+            let tmp_data_path = PathBuf::from(&tmp_extract_path).join("data.tar.gz");
+            let tmp_filesmap_path = PathBuf::from(&tmp_extract_path).join("files.map");
+
+            // Open the Manifest to retrieve the package's metadata.
+            let manifest = open_manifest(&tmp_manifest_path)?;
+
+            // Find all the files within `data.tar.gz` and write their path in `tmp_filesmap_path`.
+            if manifest.kind() == Kind::Effective {
+                let mut files = Vec::new();
+                let data_file = File::open(tmp_data_path)?;
+                let mut data = Archive::new(GzDecoder::new(data_file));
+
+                for entry in data.entries()? {
+                    let entry = entry?;
+                    let entry_path = entry.path()?;
+
+                    // Quick beautifier without altering the meaning of the path
+                    // before storing and printing it.
+                    let mut pretty_path = PathBuf::from("/");
+
+                    for component in entry_path.components() {
+                        match component {
+                            Component::CurDir => (),
+                            _ => pretty_path.push(component),
+                        }
+                    }
+                    files.push(pretty_path.display().to_string());
+                }
+
+                let filesmap = File::create(&tmp_filesmap_path)?;
+                serde_json::to_writer(filesmap, &files)?;
+            }
+
+            // If this is a reupload of an already existing package, remove the previous content
+            if self.exists() {
+                self.purge()?;
+            }
+
+            // Copy all new files to their destination
+            // We do a copy and not a move because they may not share the same mountpoint
+
+            fs::create_dir_all(&self.cache_path)?;
+            fs::copy(&tmp_manifest_path, &self.manifest_path)?;
+            fs::copy(&tmp_filesmap_path, &self.filesmap_path)?;
+        };
+        fs::remove_dir_all(&tmp_extract_path)?;
+        res
+    }
+
+    /// Return the manifest of the package.
+    pub fn manifest(&self) -> Result<Manifest, Error> {
+        open_manifest(&self.manifest_path)
+    }
+
+    /// Return, if it exists, a list of all the content present in the package.
+    pub fn filesmap(&self) -> Result<Option<Vec<String>>, Error> {
```

Comment:
Is there a reason why you return a vector of strings, instead of paths ? The latter would seem more appropriate to me

Reply:
I actually thought PathBuf didn't implement Serialize/Deserialize, I don't know why :o 

Thanks!